src/directives/NgGrid.ts

```diff
@@ -1013,36 +1044,91 @@ export class NgGrid implements OnInit, DoCheck, OnDestroy {
 	}
 
 	private _fixGridPosition(pos: NgGridItemPosition, dims: NgGridItemSize): NgGridItemPosition {
-		while (this._hasGridCollision(pos, dims) || !this._isWithinBounds(pos, dims, true)) {
-			if (this._hasGridCollision(pos, dims)) {
-				const collisions: NgGridItem[] = this._getCollisions(pos, dims);
+		if (!this._hasGridCollision(pos, dims)) return pos;
+
+		const maxRow = this._maxRows === 0 ? this._getMaxRow() : this._maxRows;
+		const maxCol = this._maxCols === 0 ? this._getMaxCol() : this._maxCols;
+		if (this._itemFixDirection === "vertical") {
+			fixLoop:
+			for (; pos.col <= maxRow;) {
+				const itemsInPath = this._getItemsInVerticalPath(pos, dims, pos.row);
+				let nextRow = pos.row;
+
+				for (let item of itemsInPath) {
+					if (item.row - nextRow >= dims.y) {
+						pos.row = nextRow;
+						break fixLoop;
+					}
 
-				switch (this.cascade) {
-					case 'up':
-					case 'down':
-					default:
-						pos.row = Math.max.apply(null, collisions.map((item: NgGridItem) => item.row + item.sizey));
-						break;
-					case 'left':
-					case 'right':
-						pos.col = Math.max.apply(null, collisions.map((item: NgGridItem) => item.col + item.sizex));
-						break;
+					nextRow = item.row + item.sizey;
 				}
-			}
 
+				if (maxRow - nextRow >= dims.y) {
+					pos.row = nextRow;
+					break fixLoop;
+				}
 
-			if (!this._isWithinBoundsY(pos, dims)) {
-				pos.col++;
+				pos.col = Math.max(pos.col + 1, Math.min.apply(Math, itemsInPath.map((item) => item.col + dims.x)));
 				pos.row = 1;
 			}
-			if (!this._isWithinBoundsX(pos, dims)) {
-				pos.row++;
+		} else if (this._itemFixDirection === "horizontal") {
+			fixLoop:

```

Comment:
I'm not a fan of this, but if it gets the job done I can't complain :smile:

Reply:
What would you propose instead?